landscaper/gardenlet/pkg/gardenlet_delete.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pkg
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/utils/retry"
+
+	"github.com/gardener/gardener/landscaper/gardenlet/pkg/applier"
+)
+
+// Delete removes all deployed Gardenlet resources from the Seed cluster.
+func (g Landscaper) Delete(ctx context.Context) error {
+	shootList := &gardencorev1beta1.ShootList{}
+	err := g.gardenClient.Client().List(ctx, shootList)
+	if err != nil {
+		return fmt.Errorf("cannot delete Seed '%s'. Failed to check if Seed is still in use by shoot(s): %v", g.gardenletConfiguration.SeedConfig.Name, err)
```

Comment:
```suggestion
		return fmt.Errorf("failed to list shoots: %w", g.gardenletConfiguration.SeedConfig.Name, err)
```
The wrapping error message should explain what actually failed. "cannot delete seed" is not appropriate here, instead the caller of the `Delete` method could add it if it wants to.
I recently learned what `%w` does with `fmt.Errorf`, and consider it to be a best practice, see https://blog.golang.org/go1.13-errors. I would suggest always using it when wrapping an error via `fmt.Errorf`.

Reply:
Good stuff, thanks.